### ISSUE TITLE
<string conversion failed> sometimes only with multi thread.

### ISSUE THREAD

**Issue by angelozerr** [eclipsesource/J2V8#55]
@irbull  At first I have integrated your locker API inside tern.java See https://github.com/angelozerr/tern.java/commit/261944ef12e11f83596c55bbb73cd047401eebf5#diff-62671038bbaaee6ab7ff4aa4c93bb895L76

It starts working, but I have often `<string conversion failed>` error when I open completion. I know it's not enough information but today I don't find a scenario which causes every time this problem. This problem seems occuring only in multi thread context.

I have used v8 like this : 

```java
private synchronized JsonObject request(TernDoc doc) throws TernException {
	String script = new StringBuilder("server.request(")
			.append(doc.toString()).append(");").toString();
	V8 v8 = getV8();
	try {
		v8.getLocker().acquire();
		String json = v8.executeStringScript(script);
		return (JsonObject) JsonValue.readFrom(json);
	} finally {
		v8.getLocker().release();
	}
}
```

 * I have used synchronized  for teh method. Must I do that?
 * is a correct use of locker?

If you have an idea why I have this error, it should be very cool. Thanks!

**Comment by irbull** [comment]
Do you know where <string conversion failed> comes from? Is this a Json
error? Something that is thrown from V8? Does it come with a stack trace,
or a crash report?

This use looks ok, and if you plan on constantly acquiring and releasing
the locker, then yes you will need to use synchronized.

On Wed, May 20, 2015 at 3:56 AM, Angelo <notifications@github.com> wrote:

> @irbull <https://github.com/irbull> At first I have integrated your
> locker API inside tern.java See angelozerr/tern.java@261944e
> #diff-62671038bbaaee6ab7ff4aa4c93bb895L76
> <https://github.com/angelozerr/tern.java/commit/261944ef12e11f83596c55bbb73cd047401eebf5#diff-62671038bbaaee6ab7ff4aa4c93bb895L76>
>
> It starts working, but I have often <string conversion failed> error when
> I open completion. I know it's not enough information but today I don't
> find a scenario which causes every time this problem. This problem seems
> occuring only in multi thread context.
>
> I have used v8 like this :
>
> private synchronized JsonObject request(TernDoc doc) throws TernException {
>     String script = new StringBuilder("server.request(")
>             .append(doc.toString()).append(");").toString();
>     V8 v8 = getV8();
>     try {
>         v8.getLocker().acquire();
>         String json = v8.executeStringScript(script);
>         return (JsonObject) JsonValue.readFrom(json);
>     } finally {
>         v8.getLocker().release();
>     }
> }
>
>
>    - I have used synchronized for teh method. Must I do that?
>    - is a correct use of locker?
>
> If you have an idea why I have this error, it should be very cool. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eclipsesource/J2V8/issues/55>.
>



-- 
R. Ian Bull | EclipseSource Victoria | +1 250 477 7484
http://eclipsesource.com | http://twitter.com/eclipsesource

**Comment by angelozerr** [comment]
> Do you know where <string conversion failed> comes from?

It seems that it's V8 which throws this error and I suppose it comes from https://github.com/eclipsesource/J2V8/blob/J2V8_3_0/jni/com_eclipsesource_v8_V8Impl.cpp#L35

>  Is this a Json error?

I have tried with a simple string like this : 

```java
v8.executeStringScript("'{}'") 
```

and the problem occurs too. So the problem is about V8#executeStringScript

The thing I have notices is that some thread which works and some other thread which doesn't work. The thread is created by the Eclipse JobManager. 

I have noticed, If a thread A which executes V8#executeStringScript and throw the error <string conversion failed> (which is catched), next time that this thread is used to execute V8,  V8#executeStringScript throws the error <string conversion failed> every time. If a thread B can execute  V8#executeStringScript, next time it works.

I cannot reproduce the problem with a simple Java main -( I don't understand why some thread can execute  V8#executeStringScript  and some other threads cannot? 

Don't hesitate to tell me, If you need some other infos.

Thank a lot for your help!

**Comment by irbull** [comment]
Comes from J2V8? :-) opps... I didn't even notice my own error messages :-).

I have tern.java cloned and all the projects checked out. I've tried the SWT example, but as you said, it works fine. I have some compile errors because I don't have JSDT, do I need that? Any tips on launching? 

I will look into this.

**Comment by irbull** [comment]
No answers yet, but some observations. The ToCString actually comes from [1] (the V8 Sample). The idea is that if the string is not null, it converts it from a V8 String to a C String (if it's null, we get the error). The ToCString method is only used in exception handling, and it's used to covert the filename, stacktrace and message from a V8String to a CString.

That all makes sense. But why those values would be NULL, and why only with some threads is a mystery. 

[1] https://v8.googlecode.com/svn/trunk/samples/shell.cc

**Comment by irbull** [comment]
A few more questions:
1. Is tern.TernException a java exception?
2. Is this re-entrant? (is it Java that's throwing the exception, after executing a script, which in turn comes from a script?)

**Comment by irbull** [comment]
I can somewhat reproduce this if I do the following

1. Register a Java callback from JS
2. In that callback, release the runtime, create a new thread, access the runtime

However, in this case we now have to different threads using the same V8 Isolate at the same time (since the isolate will be in use from 1 and then accessed by a different thread in 2). V8 does not allow this and the results will be unpredictable. 

If you really need to use a second thread in callback, you should create a new V8Runtime.

**Comment by angelozerr** [comment]
Sorry @irbull for the delay. Here my answers : 

> I have tern.java cloned and all the projects checked out. I've tried the SWT example, but as you said, it works fine. 

Yes it works because SWT example uses tern with "classic" completion (no multi thread).

> I have some compile errors because I don't have JSDT, do I need that? Any tips on launching?

tern.java can be used with simple SWT like you have seen and was integrated inside Eclipse IDE. JSDT completion was extended
with a JSDT extension point to use tern. When you have a lot of JavaScript files, ternjs can takes some time.
So to avoid freezing Eclipse IDE, the tern completion is executed with a Job and if it takes too time, the completion displays 
a item completion which tell that tern is computing.

It is this case which doesn't work with J2V8 : use tern.java inside Eclipse IDE.

So to test J2V8 inside Eclipse IDE, you need to : 

 * download "Eclipse JEE" like Luna. You will have already installed JSDT. After that you will able to start tern.java inside Eclipse IDE.
 * launch the tern.java plugins to have a seconf Eclipse IDE which load tern.java
 * Go at Window -> Preferences -> Tern and select "Tern with J2V8" as "Tern server type" because by default node.js is used.
 * create a Simple project (or JavaScript project if you wish).
 * Select your project, click on mouse right button to open contextual menu and click on "Configure -> Convert to Tern project..."
 * this action will create a .tern-project in your Eclipse project and will open Tern project properties.
 * go at Tern -> Modules and selct for instance tabris to benefit with tabris completion.
 * create an empty JavaScript file and tries to open completion (with Ctrl+Space). The first time it should work, but if you open completion several times
  you should have the <string conversion failed> error.

Please get the last tern.java sources (I have just commited clean code and add synchronized).

> 1. Is tern.TernException a java exception?

Yes, but it is never thrown inside JavaScript.

> 2. Is this re-entrant? (is it Java that's throwing the exception, after executing a script, which in turn comes from a script?)

No, it'not that. tern.TernException is thrown for instance if V8 fails.

> That all makes sense. But why those values would be NULL, and why only with some threads is a mystery.

I don't know -(

> I can somewhat reproduce this if I do the following

> 1. Register a Java callback from JS
> 2. In that callback, release the runtime, create a new thread, access the runtime

The only Java Callback that I have registered with v8.registerJavaMethod is to use console.log inside the JS (and display trace in the Eclipse console). But even if I remove it, the error occurs.

Perhaps I'm wrong, but I don't think I have this case. I have this case : 

 * the first time when completion is executed a Thread A create the V8 instance and uses this V8 instance to execute completion.
 * the other time, Thread B, C or A uses the V8 instance (created by A) to execute completion. A works every time. B doesn't work, C works (what is the difference between B and C?)

Perhaps it is the problem that Thread A creates the V8, but which thread must be created the V8? I don't manage thread, it's the Job Eclipse which does that.

And even if I execute a simple script like v8.executeStringScript("'a'"); the error occurs.

Hope my comments will help you.

**Comment by irbull** [comment]
I just got up and haven't tested this yet. But I wonder if get getV8()
needs to be synchronized? Is there a race condition where two threads are
trying to create two instances of v8? I will follow your steps and try to
launch this myself later today, but first I need coffee :-).

On Thu, May 21, 2015 at 1:19 AM, Angelo <notifications@github.com> wrote:

> Sorry @irbull <https://github.com/irbull> for the delay. Here my answers
> :
>
> I have tern.java cloned and all the projects checked out. I've tried the
> SWT example, but as you said, it works fine.
>
> Yes it works because SWT example uses tern with "classic" completion (no
> multi thread).
>
> I have some compile errors because I don't have JSDT, do I need that? Any
> tips on launching?
>
> tern.java can be used with simple SWT like you have seen and was
> integrated inside Eclipse IDE. JSDT completion was extended
> with a JSDT extension point to use tern. When you have a lot of JavaScript
> files, ternjs can takes some time.
> So to avoid freezing Eclipse IDE, the tern completion is executed with a
> Job and if it takes too time, the completion displays
> a item completion which tell that tern is computing.
>
> It is this case which doesn't work with J2V8 : use tern.java inside
> Eclipse IDE.
>
> So to test J2V8 inside Eclipse IDE, you need to :
>
>    - download "Eclipse JEE" like Luna. You will have already installed
>    JSDT. After that you will able to start tern.java inside Eclipse IDE.
>    - launch the tern.java plugins to have a seconf Eclipse IDE which load
>    tern.java
>    - Go at Window -> Preferences -> Tern and select "Tern with J2V8" as
>    "Tern server type" because by default node.js is used.
>    - create a Simple project (or JavaScript project if you wish).
>    - Select your project, click on mouse right button to open contextual
>    menu and click on "Configure -> Convert to Tern project..."
>    - this action will create a .tern-project in your Eclipse project and
>    will open Tern project properties.
>    - go at Tern -> Modules and selct for instance tabris to benefit with
>    tabris completion.
>    - create an empty JavaScript file and tries to open completion (with
>    Ctrl+Space). The first time it should work, but if you open completion
>    several times you should have the error.
>
> Please get the last tern.java sources (I have just commited clean code and
> add synchronized).
>
>
>    1. Is tern.TernException a java exception?
>
>  Yes, but it is never thrown inside JavaScript.
>
>
>    1. Is this re-entrant? (is it Java that's throwing the exception,
>    after executing a script, which in turn comes from a script?)
>
>  No, it'not that. tern.TernException is thrown for instance if V8 fails.
>
> That all makes sense. But why those values would be NULL, and why only
> with some threads is a mystery.
>
> I don't know -(
>
> I can somewhat reproduce this if I do the following
>
>    1. Register a Java callback from JS
>    2. In that callback, release the runtime, create a new thread, access
>    the runtime
>
>  The only Java Callback that I have registered with v8.registerJavaMethod
> is to use console.log inside the JS (and display trace in the Eclipse
> console). But even if I remove it, the error occurs.
>
> Perhaps I'm wrong, but I don't think I have this case. I have this case :
>
>    - the first time when completion is executed a Thread A create the V8
>    instance and uses this V8 instance to execute completion.
>    - the other time, Thread B, C or A uses the V8 instance (created by A)
>    to execute completion. A works every time. B doesn't work, C works (what is
>    the difference between B and C?)
>
> Perhaps it is the problem that Thread A creates the V8, but which thread
> must be created the V8? I don't manage thread, it's the Job Eclipse which
> does that.
>
> And even if I execute a simple script like v8.executeStringScript("'a'");
> the error occurs.
>
> Hope my comments will help you.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eclipsesource/J2V8/issues/55#issuecomment-104174863>.
>



-- 
R. Ian Bull | EclipseSource Victoria | +1 250 477 7484
http://eclipsesource.com | http://twitter.com/eclipsesource

**Comment by angelozerr** [comment]
> I just got up and haven't tested this yet. But I wonder if get getV8() needs to be synchronized?

No, I think not since request method is synchronized.

>  I will follow your steps and try to launch this myself later today, 

Thanks!

> but first I need coffee :-).

Good luck an dmany thanks again!

**Comment by angelozerr** [comment]
And just for your info I'm on Windows 64 bits.

**Comment by irbull** [comment]
I'm on MacOS, and while I can fire up a windows VM I thought I would try in
on Mac. I get similar errors (although not exactly the same).

One thing I did and it seemed to fix it is I released the v8 runtime in a
finally block of the getV8 method. When you create the runtime, it's also
acquired. It's then re-acquired in the request method and released there.

```
private V8 getV8() throws TernException {
if (v8 == null) {
v8 = V8.createV8Runtime();
loadTern();
try {
v8.registerJavaMethod(this, "log", "_javaConsole",
new Class<?>[] { String.class, Integer.class });
v8.executeVoidScript(IOUtils.toString(J2V8TernServer.class
.getResourceAsStream("tern-j2v8.js")), "tern-j2v8.js",
0);

StringBuilder script = new StringBuilder(
"var server = new J2V8TernServer(");
script.append("[");
script.append(defs.toString());
script.append("],");
script.append(getProject().getPlugins() != null ? getProject()
.getPlugins().toString() : "");
script.append(");");
v8.executeVoidScript(script.toString());
} catch (IOException e) {
throw new TernException(e);
} finally {
v8.getLocker().release();
}
}
return v8;
}
```

Can you try this on windows?

On Thu, May 21, 2015 at 6:33 AM, Angelo <notifications@github.com> wrote:

> And just for your info I'm on Windows 64 bits.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eclipsesource/J2V8/issues/55#issuecomment-104278324>.
>



-- 
R. Ian Bull | EclipseSource Victoria | +1 250 477 7484
http://eclipsesource.com | http://twitter.com/eclipsesource

**Comment by irbull** [comment]
I have spent most of the day on this, and while I can easily reproduce this problem (even on MacOS), I honestly have no idea :-(. I've changed the logic in J2V8, and I can demonstrate that
1. An exception exists
2. It's not empty
3. It has a message (that's not empty)
4. But when I convert it to a String I get NULL

The only think I can think of is the comments in TryCatch in V8.h
```
  /**
   * Creates a new try/catch block and registers it with v8.  Note that
   * all TryCatch blocks should be stack allocated because the memory
   * location itself is compared against JavaScript try/catch blocks.
   */
```
I wonder if because we execute the JS different threads, if the mapping from try/catch in JS to V8 get messed up and memory location comparison is failing.

Why it only happens with Jobs (and not just threads) is a mystery. Also, I extended Tern to create a thread (inside) the job, and that worked fine.

There are a few solutions we could use in tern:
1. Use a V8Runtime per thread (get the Runtime for the thread, if it doesn't exist, create it).
2. Use a single thread and sync it
3. Create a new V8Runtime each time (I did this, but it's a bit slow).

**Comment by irbull** [comment]
I take that back, I may have a solution.

Of course, you work on a problem for 12 hours, give up, eat some icecream, and the solution is obvious :-).

I made the same assumption that Clemens made here [1], that is, I assumed if I used my own locking then I didn't need the V8::Locker. But that's not true. When I added a V8Locker, things started to work. I need to work out the details, but I think we have a path forward. 

[1] https://groups.google.com/forum/#!topic/v8-users/5Kdqlpt0NxU

Sorry for all the spam, but thanks for the bug report. This was a good one.

**Comment by angelozerr** [comment]
Thank's @irbull for your support and good luck!

**Comment by irbull** [comment]
Fixed with 8d30673f64a8a6a773635f7a502420d5c1de8c7f. I'll let you know when it's in maven central.

**Comment by angelozerr** [comment]
@irbull I have getted your last 3.0.0 SNAPSHOT and play with it inside tern.java and it seems that the error with multi threading doesn't occurs more.

Very cool, Many thanks!
